add integrations page for chat docs (#2940)
ably/docs: src/pages/docs/chat/integrations.mdx

| `version.timestamp` | `version.timestamp` | Use message `timestamp` if no `version` is set. |
| key `"reaction:unique.v1"` in `annotations.summary` | `reactions.unique` | Message reactions of type unique |
| key `"reaction:distinct.v1"` in `annotations.summary` | `reactions.distinct` | Message reactions of type distinct |
| key `"reaction:multiple.v1"` in `annotations.summary` | `reactions.multiple` | Message reactions of type multiple |

If you are using a language that has an Ably Pub/Sub SDK, you can use the `fromEncoded()` or `fromEncodedArray()` methods to decode the encoded message (the JSON or msgpack received) into a Pub/Sub message. Then you can apply the mapping from the table above to get the Chat message form the Pub/Sub message.

If you are not using a language supported by an Ably Pub/Sub SDK, you will need to manually decode the message.

Be aware that message actions are numeric values over the realtime channels, Pub/Sub REST API, and in integrations. Ably SDKs convert these numeric values to the corresponding string representations. If you do not use an Ably SDK, you will need to convert the numerical action values to their string representations.

> If possible, use an Ably Pub/Sub SDK with `fromEncoded()` or `fromEncodedArray()` to decode the message. This will ensure that the message is decoded correctly and all the fields are present, as well as convert things like numerical action values to their string representations. Details on how to do this are available in the [platform integrations documentation](/docs/platform/integrations/webhooks).

Reviewer: This feels very close to the paragraph directly under the table - perhaps axe one?

---
title: Integrations
meta_description: "Ably Chat integrations with external services."
---

Ably Chat room use Ably Pub/Sub channels as the underlying building block. Each chat room uses one Pub/Sub channel. This allows you to integrate with external services using any of the [integrations](/docs/platform/integrations) that Ably Pub/Sub supports, enabling you to send data from Ably to an external service or push data into Ably from an external service.

> To integrate with an external service, you need to be aware of how Ably Chat messages are mapped to Ably Pub/Sub messages. All the Chat SDKs automatically do this conversion for you, and the Ably Chat REST API directly returns Chat messages. However, there are some situations, such as building integrations we don't yet support directly in Chat, where understanding the mapping from an Ably Pub/Sub message to a Chat message is useful.

Reviewer: ```suggestion
To integrate with an external service, you need to be aware of how Ably Chat messages are mapped to Ably Pub/Sub messages. All the Chat SDKs automatically do this conversion for you, and the Ably Chat REST API directly returns Chat messages. However there are some situations, such as building integrations, that we don't yet support directly in Chat, where understanding the mapping from an Ably Pub/Sub message to a Chat message is useful.
```

If you're looking to add moderation, Ably Chat already has [built-in moderation](/docs/chat/moderation) support for Hive, Tisane, and Bodyguard, and you can also [build custom moderation](/docs/chat/moderation/custom).

> This guide is also applicable if you are using Ably Chat with a language or platform that doesn't yet have an Ably Chat SDK, but it has an Ably Pub/Sub SDK.

Reviewer: ```suggestion
This guide is also applicable if you are using Ably Chat with a language or platform that doesn't yet have an Ably Chat SDK, but does have an Ably Pub/Sub SDK.
```

|-------------------|-------------------|---------------------|
| `clientId` | `clientId` | Direct mapping with empty string default |
| `timestamp` | `timestamp` | In Pub/Sub it's a unix timestamp in milliseconds since the epoch. In Chat it's a Date object. |
| `serial` | `serial` | Direct mapping with empty string default |
| `data.text` | `text` | Data must be JSON-decoded first. |
| `data.metadata` | `metadata` | Data must be JSON-decoded first. |
| `extras.headers` | `headers` | Flat object, key-value pairs. |
| `action` | `action` | `message.create`, `message.update`, `message.delete`, or `message.summary` |
| `version.serial` | `version.serial` | Use message `serial` if no `version` is set. |
| `version.timestamp` | `version.timestamp` | Use message `timestamp` if no `version` is set. |
| key `"reaction:unique.v1"` in `annotations.summary` | `reactions.unique` | Message reactions of type unique |
| key `"reaction:distinct.v1"` in `annotations.summary` | `reactions.distinct` | Message reactions of type distinct |
| key `"reaction:multiple.v1"` in `annotations.summary` | `reactions.multiple` | Message reactions of type multiple |

> If you are using a language that has an Ably Pub/Sub SDK, you can use the `fromEncoded()` or `fromEncodedArray()` methods to decode the encoded message (the JSON or msgpack received) into a Pub/Sub message. Then you can apply the mapping from the table above to get the Chat message form the Pub/Sub message.

Author: There's a link near this paragraph now that points to a page with a code example.

I could duplicate it but maybe it's good as it is?

- [Message actions](https://sdk.ably.com/builds/ably/specification/main/features/#TM5) are numeric values over the realtime channels, Pub/Sub REST API, and in integrations. Ably SDKs convert these numeric values to the corresponding string representations. If you do not use an Ably SDK, you will need to convert the numerical action values to their string representations.
- If possible, use an Ably Pub/Sub SDK with `fromEncoded()` or `fromEncodedArray()` to decode the message. This will ensure that the message is decoded correctly and all the fields are present, as well as convert things like numerical action values to their string representations. Details on how to do this are available in the [platform integrations documentation](/docs/platform/integrations/webhooks).
- Message reaction summaries are always sent with the message, so you can use the `annotations.summary` field to get the reaction summaries, even if the action is not `message.summary`.

> ### Is this message new, updated, or deleted?

Author: done, I've moved things around a bit and got rid of some repetition.
